### ISSUE TITLE
vertx-web - Have a way of extracting client .js files to webroot folder

### ISSUE THREAD

**Issue by davsclaus** [fabric8io/vertx-maven-plugin#21]
When doing a vertx-web you may use vertx javascript .js files in html files. And today you need to extract the .js file and copy those to your webroot folder in your source code.

I would like to have some way of configuring / auto detecting this from this maven plugin.

Detect

- vertx-web as dependency in pom.xml
- webroot in the src/main/resources directory

Then extract the .js files from vertx-client to the fat-jar in the webroot folder. We should have a configuration to allow sub folder, as sometimes people configure static files in `webroot/asserts` or something.

Here is a little example
https://github.com/camelinaction/camelinaction2/tree/master/chapter7/vertx/src/main/resources/webroot

Sadly vertx guys are not so "easy of use" focused. And webjars dont work. They drag in extra dependencies that are not in maven central. And its also sad to use dependencies from outside io.vertx. It should be simpler.

 But the client files are in .js files you can find on maven central at

https://repo1.maven.org/maven2/io/vertx/vertx-web/3.3.3/vertx-web-3.3.3-client.js

So we can make the plugin download that JAR and include those 2 JS files in the webroot folder.

The sweet spot for vert.x is those HTML5 reactive web apps, and it should be much easier to get started code and develop.

**Comment by pmlopes** [comment]
webjars.org does work:

```xml
<dependency>
    <groupId>org.webjars.bower</groupId>
    <artifactId>vertx3-eventbus-client</artifactId>
    <version>3.3.0</version>
</dependency>
```

**Comment by davsclaus** [comment]
We could also consider having the plugin copy the .js files into the src/main/resources/webroot folder so they can be there for developer editor to "see" so when the user edits the html files then the editor can locate the .js file and provide assistance.

**Comment by davsclaus** [comment]
Okay but now you have version 3.3.0 in webjars, and use vert.x version 3.3.3. A bit confusing for new users.

**Comment by rhuss** [comment]
There is a 

```xml
<dependency>
    <groupId>org.webjars.npm</groupId>
    <artifactId>vertx3-eventbus-client</artifactId>
    <version>3.3.3</version>
</dependency>
```
Tbh I dont find it too appealing to add an extra dependency resolution in addition to Maven's. E.g there will be situation where everything is proxied over a Nexus or so, with no external internet access allowed. The the custom download would fail, but webjars do still work. Also, what happens if someone adds a webjar in addition of the implicit download mechanism ?

BTW, would you download the javascript into the working dir `target/` or into the source dir `src/` ? If it is uncoditional I only would do it into `target/` to avoid overwriting source code. For everything below `src/` the user should have the complete control. (Thats also why generated code is put into a different directory).

**Comment by kameshsampath** [comment]
my take would be to have a specialized goal like springboot and swarm has like `setup` which can do all these kind of tasks .. though for now its only adding js tomorrow something might come up as well

**Comment by rhuss** [comment]
Sounds like a good compromise 👍

**Comment by davsclaus** [comment]
Yeah the setup goals is better. Let me close this ticket then some day we may work on setup goals, if the vert.x team dont realize that new users need help to get started

**Comment by cescoffier** [comment]
@davsclaus, @rhuss  Check https://github.com/fabric8io/vertx-maven-plugin/pull/28, and the web example using Bower to resolve JS dependencies, and Grunt to do some frontend tasks (it also demonstrates how you can use any plugin replayed during redeploy).

Of course, the pom.xml file is not very clear here, because mixing frontend tools with Maven is not straightforward, but it works and inherits from the whole NPM/Gulp/Bower/Grunt/WebPack/_the next cool frontend technology_... ecosystem.

The example is there: https://github.com/cescoffier/vertx-maven-plugin-1/tree/master/samples/web-template-example

**Comment by kameshsampath** [comment]
Thanks to @cescoffier , its been resolved via #32
